### PR TITLE
kernel: sched: Change cpu pin only for not executing threads 

### PR DESCRIPTION
**Pull request by ceolin**
Do not allow changing the CPU which a thread is pinned when it is
already being executed. This possibilities further optimizations in
some platforms with incoherent memory since we can safely assume that
the thread will run in the same CPU and avoid invalidate / flush the
cache during context switches.

### CODE REVIEW

File: kernel/sched.c [hunk 1]
```diff
@@ -1594,6 +1594,12 @@ static int cpu_mask_mod(k_tid_t thread, uint32_t enable_mask, uint32_t disable_m
 {
 	int ret = 0;
 
+#if defined(CONFIG_ASSERT) && defined(CONFIG_SCHED_CPU_MASK_PIN_ONLY)
```
Comment (marc-hb) [inline kernel/sched.c:1597]:
```suggestion
#if defined(CONFIG_SCHED_CPU_MASK_PIN_ONLY)
```
https://docs.zephyrproject.org/latest/kernel/services/other/fatal.html

**Review by marc-hb — COMMENTED** [review]
> This possibilities further optimizations in some platforms with incoherent memory...

syntax error or new verb "to possibility"?

File: kernel/sched.c (hunk 1 above)
Comment (ceolin) [reply kernel/sched.c:1597]: I thought about it, but that is the pattern used few line bellow

**Comment by ceolin** [comment]
> > This possibilities further optimizations in some platforms with incoherent memory...
> 
> syntax error or new verb "to possibility"?

Ignorance from my part :)

File: kernel/Kconfig [hunk 2]
```diff
@@ -148,7 +148,8 @@ config SCHED_CPU_MASK_PIN_ONLY
 	  architecture and want to make their own decisions about how
 	  to assign work to CPUs.  In that circumstance, some moderate
 	  optimizations can be made (e.g. having a separate run queue
-	  per CPU, keeping the list length shorter).  Most
+	  per CPU, keeping the list length shorter). Threads can be
+	  pinned only when they have not started or are dead.  Most
```
Comment (andyross) [inline kernel/Kconfig:152]: I guess I'd rephrase this to clarify, something like: "When selected, the CPU mask becomes an immutable thread attribute.  It can only be modified before a thread is started."

File: kernel/sched.c [hunk 3]
```diff
@@ -1594,6 +1594,12 @@ static int cpu_mask_mod(k_tid_t thread, uint32_t enable_mask, uint32_t disable_m
...
+	__ASSERT((thread->base.thread_state != _THREAD_DEAD) ||
+		 (thread->base.thread_state != _THREAD_PRESTART),
```
Comment (andyross) [inline kernel/sched.c:1599]: While it's true that it's safe to call this on a dead thread, it's not useful.  Thread structs that get restarted get all their state reinitialized (thus losing this setting before it can have any effect) and enter PRESTARTED anyway.

File: kernel/sched.c (hunk 1 above)
Comment (andyross) [reply kernel/sched.c:1597]: Should fix the other usage then.  The __ASSERT() macro is always defined (though potentially to an empty string), you don't need to test for anything to use it.

Comment (marc-hb) [reply kernel/sched.c:1597]: The `CONFIG_ASSERT` below is there for a reason: `int m = ...`. No such reason here.

There are dozens of other `__ASSERT`s in this file and they almost never use `CONFIG_ASSERT`.

**Review by andyross — COMMENTED** [review]
Some nitpicks.  Seems very reasonable to me, though I worry greatly that this pushes us yet farther down the road away from a properly symmetric architecture.  Hyper-tuned attention to CPU choice is more an addiction than a feature.  Apps that rely on this optimization are going to have a hard time migrating to solutions that scale naturally instead of manually.

But in the near term I don't see why this would be a problem at all.  Certainly it helps with the KERNEL_COHERENCE context switch overhead on apps like SOF that already made their own CPU decisions.

File: kernel/sched.c (hunk 3 above)
Comment (ceolin) [reply kernel/sched.c]: done, just removed the check for dead thread.

File: kernel/Kconfig (hunk 2 above)
Comment (ceolin) [reply kernel/Kconfig]: indeed. this is better. Just changed.

File: kernel/sched.c [hunk 4]
```diff
@@ -1608,7 +1608,7 @@ static int cpu_mask_mod(k_tid_t thread, uint32_t enable_mask, uint32_t disable_m
 		}
 	}
 
-#if defined(CONFIG_ASSERT) && defined(CONFIG_SCHED_CPU_MASK_PIN_ONLY)
+#ifdef CONFIG_SCHED_CPU_MASK_PIN_ONLY
```
Comment (andyross) [inline kernel/sched.c:1611]: Oh.... sorry, I didn't read far enough.  That #ifdef isn't protecting the use of __ASSERT, it's protecting the "m" value which would otherwise be flagged as an unused variable by the compiler.

Comment (ceolin) [reply kernel/sched.c:1611]: yep, fair enough. Will just remove this commit from this pr.

**Review by andyross — APPROVED** [review]
Looks great to me

**Review by dcpleung — APPROVED** [review]

**Review by teburd — APPROVED** [review]
